Expose policy info and utility methods to be available in the web-client

Eligioo opened the pull request:
## What's in this pull request?
Annotate Policy primitives including the static utility functions with `wasm-bindgen` so they can be used by the `web-client`.

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

Eligioo committed:
Expose policy info and utility methods to be available in the web-client

sisou commented on primitives/src/policy.rs:
```diff
@@ -6,6 +6,12 @@ use once_cell::sync::OnceCell;
+#[cfg_attr(
+    feature = "ts-types",
+    derive(nimiq_serde::Serialize),
+    serde(rename_all = "camelCase"),
```
It seems this rename rule has no effect, in the types the properties are still snake_cased.

sisou replied:
Sleeping over it, I noticed where our confusion may come from.

The serde attributes are only respected by `tsify`, not by `wasm-bindgen`. Unfortunately, wasm-bindgen doesn't have any rule like that ([issue still open](https://github.com/rustwasm/wasm-bindgen/issues/1818)), so you must add a `js_name` rule to all properties instead :/

sisou commented on primitives/src/policy.rs:
```diff
@@ -180,8 +186,17 @@ impl Policy {
+    }
+}
+
+#[cfg_attr(feature = "ts-types", wasm_bindgen::prelude::wasm_bindgen)]
```
Instead of having the long fully qualified import path everywhere, you can conditionally import a package only when a feature is active. That way you can use it with it's short direct name.

sisou committed:
Conditionally import wasm_bindgen

sisou committed:
Add WASM getter definitions for Policy getters

sisou committed:
Skip Policy's own fields in WASM, remove Client.policy()

The struct's own fields don't work without initialization. They are replaced by the getters from the previous commit, that do initialization internally.

Thanks to @hrxi for finding the `cfg_eval` workaround to use wasm_bindgen attributes on struct fields.

sisou commented:
I fixed a few things, now the only thing remaining is to add getters for the `pub const`s when the "ts-types" feature is enabled.

sisou committed:
Reimplement Policy's `pub const`s as getters for WASM

sisou commented:
Added the getters for the `const`s. However, now we have two types of getters. `CAPITAL_SNAKE_CASED` and `pascalCased` ones:

![image](https://github.com/nimiq/core-rs-albatross/assets/1828163/b36ea9bb-3280-4195-9c3e-07eb53fa05e8)

I am torn between:
1. Changing the `pascalCased` ones to `CAPITAL_SNAKE_CASE`, because that would be consistent with best-practices in JS/TS for static, never-changing (during runtime) variables. This renaming would only affect the JS-glue code, no renaming in Rust is required.
2. Keeping them as-is, to keep them consistent with how they are defined in Rust.

Any opinions?

jsdanielh commented on primitives/src/policy.rs:
```diff
@@ -445,16 +501,161 @@ impl Policy {
+    /// This is the address for the staking contract. Corresponds to
+    /// 'NQ77 0000 0000 0000 0000 0000 0000 0000 0001'
+    #[cfg_attr(feature = "ts-types", wasm_bindgen(getter = STAKING_CONTRACT_ADDRESS))]
+    pub fn WASM_STAKING_CONTRACT_ADDRESS() -> String {
```
Can we actually have the function name in lower case (and keeping the `wasm_bindgen` `getter` property in upper case)?

sisou replied:
Yes

sisou replied:
Done in 70eacbe9323e7c8223dde6ff57b78a86cc8fe055

sisou committed:
Rename other getters to UPPERCASE in WASM, too
